modules/tabs/main/index.coffee

```diff
@@ -1,36 +1,116 @@
-onTabSelected = (tabs, element, i) ->
+onTabSelected = (tabs, element, i, cause) ->
   tabs.selected = i
   selection = hx.select(tabs.selector)
   selection.selectAll('.hx-tab').classed('hx-tab-active', false)
   hx.select(element).classed('hx-tab-active', true)
 
   context = hx.palette.context(element)
-  hx.palette.borderContext(selection.select('.hx-tabs-content'), context)
+  tabsContent = selection.select('.hx-tabs-content')
+  hx.palette.borderContext(tabsContent, context)
 
-  hx.select(tabs.selector).selectAll('.hx-tab-content').classed('hx-tab-content-hidden', true)
-  hx.select('#' + hx.select(element).attr('data-content')).classed('hx-tab-content-hidden', false)
-  tabs.emit('change', {id: i})
+  rootSelection = hx.select(tabs.selector)
+  rootSelection.selectAll('.hx-tab-content')
+    .classed('hx-tab-content-hidden', true)
+  item = tabs.items()[i]
+  if item?
+    tabs._.options.contentRenderer tabsContent.node(), item.content
+  else
+    tabToSelectSelector = '#' + hx.select(element).attr('data-content')
+    tabToSelect = rootSelection.select(tabToSelectSelector)
+    tabToSelect.classed('hx-tab-content-hidden', false)
+  tabs.emit('change', {id: i, value: i, cause})
 
 class Tabs extends hx.EventEmitter
 
-  constructor: (@selector) ->
+  constructor: (@selector, options) ->
     super
 
+    defaultRenderer = (node, value) -> hx.select(node).text(value)
+
+    defaultOpts = {
+      items: []
+      titleRenderer: defaultRenderer
+      contentRenderer: defaultRenderer
+    }
+
+    resolvedOptions = hx.merge defaultOpts, options
+
+    @_ = {}
+
+    @_.options = resolvedOptions
+
     hx.component.register(@selector, this)
 
+    rootSel = hx.select @selector
+
+    rootSel.classed 'hx-tabs', true
+
+
+    titleRenderer = resolvedOptions.titleRenderer
+
+    tabsContent = rootSel.select '.hx-tabs-content'
+
+    if tabsContent.empty()
+      tabsContent = rootSel.append 'div'
+        .class 'hx-tabs-content'
```

Comment:
ideally this would always be created in the constructor. Detecting if it exists is mostly for backwards compatibility